File: medium/Triangle.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */

/*
    思路 : 

    這一題也是 2D-DP , 要計算走到底的最小path sum 相當於
    min( 最底層的cost + 可以走到該cost的位置的最小path sum )

    這一題比較特別的反而是他的形狀 , 是一個三角形 , 不過這不影響使用2D DP,
    
    dp[i][j] 代表走到該 (i,j)位置的最小path sum , 解在最後一層的minimum 
    
    dp[i][j] = triangle[i][j] + min(   dp[i-1][j]  , dp[i-1][j-1] ) 

*/

var minimumTotal = function(triangle) {
    

    // dp table的size等於 三角形層數 x 最後一層大小 , 但實際上層數=最後一層大小 
    let size = triangle.length ; 

    let dp = Array.from( new Array(size) , (v)=> new Array(size) ) ;

    dp[0][0] = triangle[0][0] ; 
    // 外層loop推進層數往下 , 實際上traverse的方向就在於我們的dp-table中的值需要什麼方向的值優先被算好 
    // 對於這一題來說就很基本的,只要正上與左上  

    for (let i = 1  ; i < size ; i++ ){
        // 注意內迴圈的中止條件 , 我們這邊是針對左下三角的範圍去traverse 
        
        for (let j = 0 ; j <= i ; j ++ ){
            if (j==0){
                dp[i][j] = triangle[i][j] + dp[i-1][j] 
                }             
            else if (j == i){
                dp[i][j] = triangle[i][j] + dp[i-1][j-1] 
            }
            else {
                dp[i][j] = triangle[i][j] + Math.min( dp[i-1][j] ,  dp[i-1][j-1] )
            }
        }
    }
    return Math.min( ...dp[size-1]  )
};

let test_case = [[2],[3,4],[6,5,7],[4,1,8,3]] ; 
minimumTotal(test_case) ;

File: TOP150/Triangle.py
from typing import List 

# 標準的遞推 , 注意三角形 dp-table的初始化
class Solution:
    # dp solution 
    # base case : dp[0][0] = triangle[0][0] 
    # state transition : dp[i][j] = min( dp[i-1][j-1] , dp[i-1][j] ) beside two end-point
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        
        
        high = len(triangle)  
        dp = [ [ None for j in range(i) ] for i in range( 1 , high+ 1) ] 

        # base case 
        dp[0][0] = triangle[0][0]
        
        # Outer-loop traverse the vertical direction
        for i in range(1 , high):  
            
            # inner loop traverse the horizontal direction
            for j in range( 0 , i+1 ): 
                
                if j == 0 : 
                    dp[i][j] = dp[i-1][j] + triangle[i][j]
                elif j == i : 
                    dp[i][j] = dp[i-1][j-1] + triangle[i][j]
                else : 
                    dp[i][j] = min( dp[i-1][j-1] , dp[i-1][j]  ) + triangle[i][j]
        
        
        return min( dp[high-1] )
        

# 狀態壓縮 , 只保留前一列 , 增加space性能
class Solution:
    
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        
        prev_state = triangle[0] 
        high = len(triangle)
        
        for i in range(1 , high):  
            
            new_state = [] 
            
            for j in range(0  , i+1) :  
                
                if j == 0 :  
                    new_state.append( prev_state[0] + triangle[i][j] )
                elif j == i  : 
                    new_state.append( prev_state[j-1] + triangle[i][j] )
                else : 
                    new_state.append( 
                        min( prev_state[j-1] , prev_state[j]  ) + triangle[i][j]
                    )
            
            prev_state = new_state 
        
        return min(prev_state) 
    
    
# Top down + memo 

class Solution:
    
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        
        high = len(triangle) 
        memo = dict() 
        
        # dp(i,j) 的定義為到達 i , j 格的最小cost 

        def dp(i,j):  
            # base-case , i==0,j==0     
            if i == 0 and j == 0 : return triangle[0][0]
            elif (i,j) in memo : return memo[(i,j)] 
            else : 
                
                # 特殊的case在於 j==0 或著 j == i , 這樣只能往單一方向找 
                if j == 0 : 
                    
                    sol = triangle[i][j] + dp(i-1 , 0) 
                
                elif j == i : 
                    
                    sol = triangle[i][j] + dp(i-1 , j-1)
                    
                else : 
                    
                    sol = triangle[i][j] + min(
                        dp(i-1 , j-1) , dp(i-1,j)
                    )     
                
                memo[(i,j)] = sol 
                return sol 
            
        best = float('inf')
        for j in range(high):  
            best = min(best  , dp( high-1 , j  )  )
        
        return best 
                    
    